### ISSUE TITLE
Graphene support is required for these test function present in LTP suite

### ISSUE THREAD

**Issue by anjalirai-intel** [gramineproject/graphene#2297]
## Description of the problem
Missing Graphene Support to test the following testfile.

## TestFile 1: open10 (UPDATE: unfixable)
Description: Uses getpwnam("nobody") and requires "nobody" user to exist.

## TestFile 2: open14 (UPDATE: Requires new feature)
Description: It requires O_TMPFILE support

## TestFile : openat03 (UPDATE: Requires new feature)
Description: It requires O_TMPFILE support

## TestFile 4: pipe2_04 (UPDATE: partly fixed by #2524 but generally unfixable)
Description: fails to set O_NONBLOCK by using pipe2(fds, O_NONBLOCK)

**Comment by dimakuv** [comment]
> Description: fails to set O_NONBLOCK by using pipe2(fds, O_NONBLOCK)

I'm curious why this happens. We have non-blocking pipes in Graphene... Some corner case, I guess.

**Comment by vijaydhanraj** [comment]
> Description: fails to set O_NONBLOCK by using pipe2(fds, O_NONBLOCK)

It looks like we don't save the `O_NONBLOCK` flag for the write end of the pipe `fd[1]`. We simply set the `O_NONBLOCK` flag for the `PAL_HANDLE` but this isn't propagated to the `SHIM_HANDLE` when creating pipe. So, when the test tries to verify the `O_NONBLOCK` flag by `fcntl(p[1], F_GETFL)`, graphene looks up the flag from `SHIM_HANDLE` which doesn't have the flag set resulting in the error.

Created a simple fix as below,
```
diff --git a/LibOS/shim/src/sys/shim_pipe.c b/LibOS/shim/src/sys/shim_pipe.c
index ff86f209..b10db98d 100644
--- a/LibOS/shim/src/sys/shim_pipe.c
+++ b/LibOS/shim/src/sys/shim_pipe.c
@@ -53,13 +53,16 @@ static int create_pipes(struct shim_handle* srv, struct shim_handle* cli, int fl
     srv->pal_handle = hdl1;
 
     if (flags & O_NONBLOCK) {
-        /* `cli` - `hdl2` - has this flag already set by the call to `DkStreamOpen`. */
         ret = set_handle_nonblocking(srv, /*on=*/true);
         if (ret < 0) {
             /* Restore original handle, if any. */
             srv->pal_handle = tmp;
             goto out;
         }
+
+        ret = set_handle_nonblocking(cli, /*on=*/true);
+        if (ret < 0)
+            goto out;
     }
```

The test case proceeds further but fails when the pipe size is updated using fcntl ` fcntl(fd[0], F_SETPIPE_SZ,...)`. Currently, this isn't supported in graphene and will always return `-EINVAL`
@pwmarcz @dimakuv @mkow how do we want to proceed here? Do we add this support?

**Comment by pwmarcz** [comment]
As for `F_SETPIPE_SZ`: seems pretty non-trivial (AFAIK we don't have pipe buffers in LibOS internally, we use sockets from the host), if there are no applications requiring it, there is probably no reason for change.

I agree about fixing the issue with `O_NONBLOCK`, but (on the first look, at least):
* `set_handle_nonblocking` modifies the PAL handle and calls out to host, the comment is right that this is not necessary here: it looks like you can do `cli->flags |= O_NONBLOCK` instead,
* if you fix this, I recommend adding a check (`fcntl(F_GETFL)` for both ends of pipe) to our own `pipe_nonblocking` regression test.

**Comment by vijaydhanraj** [comment]
OK, so we can leave the `F_SETPIPE_SZ` implementation for now until there is a requirement from an application.

Since we update the `cli->pal_handle` only after calling `set_handle_nonblocking(cli, /*on=*/true);` this https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/sys/shim_fcntl.c#L36 will simply return 0 due to this check, https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/fs/pipe/fs.c#L137. So, we don't end up calling the host but the flags get set https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/sys/shim_fcntl.c#L41

Felt this is more aligned with the current flow rather than just adding this flag explicitly for the client. But I am fine with either approach. Although the comment is right, I don't think it adds any value here.

Yes, will update the regression test when creating PR.

**Comment by vijaydhanraj** [comment]
>TestFile 1: open10
Description: Uses getpwnam("nobody") and requires "nobody" user to exist.

Regarding this test case, we can get past the getpwnam("nobody") by adding `/etc/passwd` as part of the allowed files. But the test goes ahead and fails when changing ownership of a test directory. Currently in graphene, `chown` syscall simply verifies if the path is valid and returns without setting the `uid` or `gid`. In addition to `chown` syscall support, this test also needs `setreuid` syscall support to be implemented.
Also, when an application opens a file with `mode` having `uid`. `gid` or `sticky bit` set, graphene seems to ignore and ends up with just `user` permission and `type` being set as the effective `mode`. Need to understand this further if it is by design. So, to enable this test we need,
1. Implement complete support for `chown` syscall.
2. Implement `setreuid` syscall.
3. Honor user requested `mode` when opening a file.

**Comment by dimakuv** [comment]
@vijaydhanraj Thanks for the analysis!

## TestFile 4: pipe2_04

- I prefer @vijaydhanraj's solution to use `set_handle_nonblocking(cli, /*on=*/true)` instead of @pwmarcz's `cli->flags |= O_NONBLOCK`. Given Vijay's walk-through of the code, `set_handle_nonblocking()` does all the required checks and only updated the shim handle's attributes (and doesn't update the PAL handle's attributes).

  I would only rearrange `set_handle_nonblocking(srv, /*on=*/true)` and `set_handle_nonblocking(cli, /*on=*/true)` in the proposed patch, so that `cli` is updated before `srv` (due to error handling on `srv`).

- Regarding `F_SETPIPE_SZ `: can we just stub it? Simply return zero instead of `-EINVAL` but not do anything? Will this allow the LTP test to proceed? And we can of course add a TODO comment.

## TestFile 1: open10

Given Vijay's analysis, it looks like this test is unfixable currently. Graphene doesn't have any notion of permissions and ownership (and it's unclear if Graphene even needs this). Even if we implement such features in Graphene, this is low priority.

**Comment by pwmarcz** [comment]
>Since we update the `cli->pal_handle` only after calling `set_handle_nonblocking(cli, /*on=*/true);` this https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/sys/shim_fcntl.c#L36 will simply return 0 due to this check, https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/fs/pipe/fs.c#L137. So, we don't end up calling the host but the flags get set https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/sys/shim_fcntl.c#L41

I didn't read so deep. Then I agree, `set_handle_nonblocking()` is cleaner and there's no reason not to use it.

**Comment by vijaydhanraj** [comment]
>Regarding `F_SETPIPE_SZ` : can we just stub it? Simply return zero instead of `-EINVAL` but not do anything? Will this allow the LTP test to proceed? And we can of course add a TODO comment.

The LTP test will still fail as the test calls `F_GETPIPE_SZ ` next to validate the pipe size that was previously set.

**Comment by vijaydhanraj** [comment]
Thanks, @dimakuv @pwmarcz for the feedback. I have created PR https://github.com/oscarlab/graphene/pull/2524 to address this issue.

**Comment by vijaydhanraj** [comment]
>TestFile 2: open14
Description: It requires O_TMPFILE support
TestFile : openat03
Description: It requires O_TMPFILE support

Regarding these test cases, yes graphene currently doesn't support the creation of an unnamed temporary regular file. We basically fail here, https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/fs/shim_namei.c#L548.
 
The error is as expected according to man page, https://man7.org/linux/man-pages/man2/open.2.html
```
EISDIR pathname refers to an existing directory, O_TMPFILE and
       one of O_WRONLY or O_RDWR were specified in flags, but
       this kernel version does not provide the O_TMPFILE
       functionality.
```

**Comment by dimakuv** [comment]
> The LTP test will still fail as the test calls F_GETPIPE_SZ next to validate the pipe size that was previously set.

How bad is it to add a new SHIM_HANDLE-only attribute for pipe size (*not* propagating this to PAL)? And set this attribute on F_SETPIPE_SZ and read this attribute on F_GETPIPE_SZ? This will be a dummy but will make applications happy. Actually, is it purely a performance optimization on Linux, or does it actually have some functional behavior?

**Comment by dimakuv** [comment]
> TestFile 2: open14
> Description: It requires O_TMPFILE support
> TestFile : openat03
> Description: It requires O_TMPFILE support

Regarding this, looks like we should wait with this implementation (looks like non-trivial). I would hope for @pwmarcz to add this at some point during his FS work, since he knows these parts the best. Or just ignore this for now, until some real-world application requires this feature.

**Comment by vijaydhanraj** [comment]
>  However, I cannot imagine any real-world application relying on this brittle assumption, so we should either ignore this at all (but then LTP test will fail, and maybe some applications will fail as well), or we should have some dummy implementation like I proposed (and then LTP test succeeds hopefully, and the applications don't panic on ENOSYS).

Implementing dummy code just for LTP doesn't seem helpful. If there is a real-world application that is gated, then we can go ahead and implement this.  This is my thought, but I don't mind implementing dummy code.

**Comment by vijaydhanraj** [comment]
> Regarding this, looks like we should wait with this implementation (looks like non-trivial). 

Yes, this is not trivial, and it needs special handling as far as I understand. We can hold off on this one.

**Comment by dimakuv** [comment]
Current summary for this issue:

- `TestFile 1: open10` requires too much unimplemented functionality in Graphene (and quite useless too: `chown`, `setreuid`, real permissions). Not gonna happen.
- `TestFile 2: open14` and `TestFile 3: openat03` require new feature `O_TMPFILE`; should be fixed when Pawel finishes FS rework hopefully.
- `TestFile 4: pipe2_04` is partially fixed by #2524 but then proceeds to hit another issue: unimplemented `F_GETPIPE_SZ` option; looks useless for Graphene so not gonna happen.
